File: core/src/main/java/dev/bee/kanjianki/core/AttributionCopy.java
package dev.bee.kanjianki.core;

import java.util.ArrayList;
import java.util.List;

public final class AttributionCopy {
    public static final String DICTIONARY_FALLBACK = "KANJIDIC2 dictionary data from EDRDG, Jiten rank data, and KanjiVG stroke data.";
    public static final String KANJIVG_FALLBACK = "KanjiVG stroke data, CC BY-SA 3.0.";

    private AttributionCopy() {
    }

    public static String dictionarySources(String generatedAt, List<Source> sources, List<String> notes) {
        if (sources == null || sources.isEmpty()) {
            return "Dictionary manifest is empty.";
        }
        List<String> lines = new ArrayList<>();
        if (!safe(generatedAt).isEmpty()) {
            lines.add("Generated: " + safe(generatedAt));
        }
        for (Source source : sources) {
            appendSource(lines, source);
        }
        appendNotes(lines, notes);
        return String.join("\n", lines).trim();
    }

    public static void appendSource(List<String> lines, Source source) {
        if (source == null) {
            return;
        }
        lines.add("");
        lines.add(firstNonEmpty(source.name, source.id));
        addSourceLine(lines, "License", source.license);
        addSourceLine(lines, "URL", source.upstreamUrl);
        addSourceLine(lines, "Source", source.sourcePath);
        addSourceLine(lines, "Fetched", source.fetchDate);
        addSourceLine(lines, "Version", firstNonEmpty(
                source.databaseVersion,
                source.version,
                source.dateOfCreation
        ));
        addSourceLine(lines, "SHA-256", source.sourceSha256);
    }

    public static void appendNotes(List<String> lines, List<String> notes) {
        if (notes == null || notes.isEmpty()) {
            return;
        }
        lines.add("");
        for (String note : notes) {
            lines.add(safe(note));
        }
    }

    private static void addSourceLine(List<String> lines, String label, String value) {
        String safeValue = safe(value);
        if (!safeValue.isEmpty()) {
            lines.add(label + ": " + safeValue);
        }
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) {
            String safeValue = safe(value);
            if (!safeValue.isEmpty()) {
                return safeValue;
            }
        }
        return "";
    }

    private static String safe(String value) {
        return value == null ? "" : value.trim();
    }

    public static final class Source {
        public final String id;
        public final String name;
        public final String license;
        public final String upstreamUrl;
        public final String sourcePath;
        public final String fetchDate;
        public final String databaseVersion;
        public final String version;
        public final String dateOfCreation;
        public final String sourceSha256;

        public Source(
                String id,
                String name,
                String license,
                String upstreamUrl,
                String sourcePath,
                String fetchDate,
                String databaseVersion,
                String version,
                String dateOfCreation,
                String sourceSha256
        ) {
            this.id = id;
            this.name = name;
            this.license = license;
            this.upstreamUrl = upstreamUrl;
            this.sourcePath = sourcePath;
            this.fetchDate = fetchDate;
            this.databaseVersion = databaseVersion;
            this.version = version;
            this.dateOfCreation = dateOfCreation;
            this.sourceSha256 = sourceSha256;
        }

    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/AttributionCopy.kt
package dev.bee.kanjianki.core

import java.util.Locale

object AttributionCopy {
    private const val JAPANESE_LANGUAGE = "ja"

    @JvmStatic
    fun dictionaryFallback(): String {
        return localizedText(
            "KANJIDIC2 dictionary data from EDRDG, Jiten rank data, and KanjiVG stroke data.",
            "KANJIDIC2の辞書データ（EDRDG、Jitenの順位データ、KanjiVGの画数データ）。",
        )
    }

    @JvmStatic
    fun kanjiVgFallback(): String {
        return localizedText(
            "KanjiVG stroke data, CC BY-SA 3.0.",
            "KanjiVGの画数データ、CC BY-SA 3.0。",
        )
    }

    @JvmStatic
    fun dictionarySources(generatedAt: String?, sources: List<Source?>?, notes: List<String?>?): String {
        if (sources.isNullOrEmpty()) {
            return localizedText("Dictionary manifest is empty.", "辞書マニフェストが空です。")
        }
        val lines = ArrayList<String>()
        val generated = safe(generatedAt)
        if (generated.isNotEmpty()) {
            lines.add("${localizedText("Generated", "生成日時")}: $generated")
        }
        for (source in sources) {
            appendSource(lines, source)
        }
        appendNotes(lines, notes)
        return lines.joinToString("\n").javaTrim()
    }

    @JvmStatic
    fun appendSource(lines: MutableList<String>, source: Source?) {
        if (source == null) {
            return
        }
        lines.add("")
        lines.add(firstNonEmpty(source.name, source.id))
        addSourceLine(lines, localizedText("License", "ライセンス"), source.license)
        addSourceLine(lines, localizedText("URL", "URL"), source.upstreamUrl)
        addSourceLine(lines, localizedText("Source", "ソース"), source.sourcePath)
        addSourceLine(lines, localizedText("Fetched", "取得日"), source.fetchDate)
        addSourceLine(
            lines,
            localizedText("Version", "バージョン"),
            firstNonEmpty(
                source.databaseVersion,
                source.version,
                source.dateOfCreation,
            ),
        )
        addSourceLine(lines, "SHA-256", source.sourceSha256)
    }

    @JvmStatic
    fun appendNotes(lines: MutableList<String>, notes: List<String?>?) {
        if (notes.isNullOrEmpty()) {
            return
        }
        lines.add("")
        for (note in notes) {
            lines.add(safe(note))
        }
    }

    private fun addSourceLine(lines: MutableList<String>, label: String, value: String?) {
        val safeValue = safe(value)
        if (safeValue.isNotEmpty()) {
            lines.add("$label: $safeValue")
        }
    }

    private fun firstNonEmpty(vararg values: String?): String {
        for (value in values) {
            val safeValue = safe(value)
            if (safeValue.isNotEmpty()) {
                return safeValue
            }
        }
        return ""
    }

    private fun safe(value: String?): String {
        return value?.javaTrim() ?: ""
    }

    private fun String.javaTrim(): String {
        return trim { it <= ' ' }
    }

    private fun localizedText(english: String, japanese: String): String {
        return if (isJapaneseLocale()) japanese else english
    }

    private fun isJapaneseLocale(): Boolean = Locale.getDefault().language == JAPANESE_LANGUAGE

    class Source(
        @JvmField val id: String?,
        @JvmField val name: String?,
        @JvmField val license: String?,
        @JvmField val upstreamUrl: String?,
        @JvmField val sourcePath: String?,
        @JvmField val fetchDate: String?,
        @JvmField val databaseVersion: String?,
        @JvmField val version: String?,
        @JvmField val dateOfCreation: String?,
        @JvmField val sourceSha256: String?,
    )
}
